### DISCUSSION TITLE
[feature request] cooperate to build a ipfs search engine pinging service standard

### DISCUSSION THREAD

**Discussion by askender** [ipfs-search/ipfs-search#197] [Ideas]
I also build a dweb search engine: https://dweb-search.on.fleek.co/
source code: https://github.com/anwen/dweb-search/issues/3

I think we should work together to figure it out does it worth to do.
If yes, we can [cooperate to build a ipfs search engine pinging service standard]. and make open search engines more useful to everyone.

## More Info
If users want to register the content metadata to more open search engine, we need a better standard.

It's a little similar to: https://stackoverflow.com/questions/10376009/how-send-to-google-ping-after-add-new-post

I will ask the author of https://github.com/ipfs-search/ipfs-search  and iscn team if they have the will to build it.

The simplest way is to do a post request:
```python
metadata = {'key1': 'value1', 'key2': 'value2'} ## see json example below
r = requests.post("https://api.open-search-01.org/post", data=metadata)
print(r.json())
```

a madadata.json example (just a draft):
```json
{
  "name": "Dweb Search Logo", 
  "description": "Logo of Dweb Search Engine. You can find more info on https://dweb-search.on.fleek.co/ or ipns://dweb-search.on.fleek.co/", 
  "image": "ipfs://QmdnxKoA3teyZgzG152eh8BffwSHN6DDBz2CEruVyqAmHB",
  "license": "CC-BY-SA",
  "license_url": "http://creativecommons.org/licenses/by-sa/4.0/",
  "url": "https://cloudflare-ipfs.com/ipfs/QmdnxKoA3teyZgzG152eh8BffwSHN6DDBz2CEruVyqAmHB",
  "filetype": "png",
  "tags": ["Dweb", "Logo"],
  "iscn": "iscn://likecoin-chain/eM0AiWxCpb5wz0KRMbitxFgFC-CzVeGdXPLe7_bNG84/1",
  "authors": [{
    "name": "Askender",
    "url": "https://askender.xyz",
    "wallet": {
      "filecoin": "f1yerglsxmiocu2pdfmvfn5sz2syk5ppetixqkmdq",
      "likecoin": "cosmos1kkvsv054ga47q8w27d38a5u46jvfh3jfkghzg6"
    }
  }]
}
```
https://ipfs.io/ipfs/QmWJBqQmWzknBxdtdDRtqgaGUiidwrPrFksHAeVW6QpDAk/metadata.json
ipfs://QmWJBqQmWzknBxdtdDRtqgaGUiidwrPrFksHAeVW6QpDAk/metadata.json

**Comment by dokterbob** [comment]
Hi @askender, we don't just have the will to work on this, we have budget as well.

# Pinning service and ping-API
Our plan is to implement the [IPFS pinning service API spec](https://ipfs.github.io/pinning-services-api-spec/), so it would be trivial for users to automatically queue items as soon as they are added to IPFS. We would simply send a reply 'as if' an item was pinned, although instead it's added to our indexing queue. 

We expect to have this running around the end of the year, when we also should have the server capacity to handle this. If you feel like assisting us in the implementation, feel free to hit me up. Although, I reckon', it would be quite trivial.

In addition, it would make a lot of sense to use PubSub to tell other nodes about new content. This would actually require some API design.

We don't currently plan to add any user-provided metadata as, without proper moderation, this is prone to abuse - our specific objective being in... being objective. ;)

# Metadata standards
HOWEVER, it could be really interesting to create some kind of metadata specification, not unlike what you proposed above. Similar to what we have been discussing before. Such a standard might comprise of:

1. A consistent container for metadata, e.g. [IPLD DAG-CBOR](https://github.com/ipld/specs/blob/master/block-layer/codecs/dag-cbor.md).
2. A consistent way for a resource to signal that it is metadata about another resource, for example:
```
{
    "metadata": { ... },
    "describing": "<IPLD Link>"
}
```
3. Agreements on how to structure such metadata, e.g. [Dublin Core](https://www.dublincore.org/specifications/dublin-core/dces/), albeit in an extensible way.
4. A consistent way of creating repositories of such metadata, e.g. (sharded) folders where names are the [CID](), linking the metadata of such content. This way, metadata can be queried directly and efficiently over HTTP.
5. A consistent way to federate repositories, e.g. creating a repository of repositories, such that they can be included in one another.
6. Eventually, a well-defined way to create indexes on such metadata, effectively providing decentralized search (in a standardized fashion).

What are your thoughts on this?

Please allow me to apologise in advance for belated replies. I only have limited time and attention for email discussions.

**Comment by askender** [comment]
> We don't currently plan to add any user-provided metadata as, without proper moderation, this is prone to abuse - our specific objective being in... being objective. ;)
Yes. current we cannot prove the creator of metamask is the creator/author of the content.  So it's not objective as your said.

Possible solution 1:
- people store metadata in the file itself, such as html <meta> tags, but many file format doesn't support that.

Possible solution 2:
- The creator of content add public keys(such as eth/filecoin wallet address) to the content file and metadata file.
- The creator of content can also update the metadata of content (such as newer ipfs/ipns content...) if they can verify tthey are the owner of the public keys.
- Users/readers/supporters/curators can donate some tokens to the creator of content. Because the content and metadata has the same public keys, the metadata can be trusted.  I think this can make it objective.
- Problem: we still don't know if the creator of the content is the real creator of thought-behind-the-content. It will be better if the creator also adds a license such as CC and claims he/she/it is the author and adds his/her/its social links.  And there'll be many third-party credit reporting agencys to verify if the publickey/wallet/creator is trustful and give scores to it.

Possible solution 3:
- a DID system that is perfect and easy to use, but it still doesn't exist as far as I know;
- Depend on the thought behind the solution 2, we can develop a simple DID system which is super easy to use.

Besides your ipfs-search and our dweb-search-engine and other dweb-project, one of my friends is working on a ipfs-data-share netdisk. All of us need an extensible metadata standard and pinging service standard(sharing between all dweb service). 
This can help authors get reasonable profit without middleman; curators can get more **dimensions** to show the contents; supporters/readers can support creators they like in a more transparent way.
Then, we may build a system/framework which is much better than reddit/medium/substack/forum/bbs/NFT/discord/matrix... at least in some aspects.

**Comment by dokterbob** [comment]
Hi @askender,

*Attribution* is yet another, very interesting problem. I am aware of at least one startup (which got bought, only to be shut down), which was trying to do this by letting artists associate a hash of their content in a blockchain. It's super interesting work and I'd like to support this, but it is way out of scope for us, right now.

Personally, I am a fan of the Unix methodology of 'doing one thing, well'. I think there's a few different things:

1. 'pinging'; letting services know that content is out there
2. metadata, self-contained or otherwise (note that Google specifically won the 'search-race' by doing automated extraction, avoiding many types of biases)
3. attribution, who made what
4. rewarding, how to reward authors

At this point, 1. is definitely in scope for us. For 2. and 3., it would be nice to be on board and we'd love to support this as soon as strong standards are developed. As for 4. this is great, but I'd rather see this built into the browser, like what Brave is doing.

**Comment by ple1n** [comment]
I would suggest supporting multiple blockchains to encourage competition. Using blockchain however has problems too, there's no way to verify that the artist is actually the author, unless we force everyone to use some blockchain which is not realistic. Even if some piece of block is indeed created by someone, no one is going to examine how many other authors are involved in the content creation. 

> third-party credit reporting agency

Establishing a committee to address the issue increases centralization, and the committee themselves might not be able to identify all plagiarism in the first place, the way of which is trying to solve a decentralized problem with a centralized entity. For example, a piece of music is published on some other platforms. Then a copycat rearranges the music, making it difficult to identify for ordinary users and the committee too. When the act of plagiarism is spotted, the copycat probably has got much reward and attention. More realistically, we won't know the guy is a copycat when he/she got enough attention. If we're going to rely on post-incident verification, the committee becomes useless.

> rewarding, how to reward authors

Rewarding would encourage plagiarism if the attribution issue is not solved, unless we treat content reposters as some kind of creator. In that case reposters might end up getting more money and attention since they are good at reposting, and as said above, even claim to be the authors when they have enough influence.

I don't think a committee or something can solve this problem. Copycat exists because of high returns, and content reposters exist because of the disconnected broken internet.
